dynamic/internal/fixup/properties.go

```diff
@@ -0,0 +1,157 @@
+	"github.com/pulumi/pulumi/sdk/v3/go/common/tokens"
+
+	"github.com/pulumi/pulumi-terraform-bridge/v3/pkg/tfbridge"
+	"github.com/pulumi/pulumi-terraform-bridge/v3/pkg/tfbridge/info"
+	shim "github.com/pulumi/pulumi-terraform-bridge/v3/pkg/tfshim"
+)
+
+// Default applies the default set of fixups to p.
+//
+// The set of fixups applied may expand over time, but it should not effect providers that
+// correctly compile in all languages.
+func Default(p *info.Provider) error {
+	return errors.Join(
+		fixPropertyConflict(p),
+		fixMissingIDs(p),
+	)
+}
+
+func fixMissingIDs(p *info.Provider) error {
+	getIDType := func(r *info.Resource) tokens.Type {
+		s := r.Fields["id"]
+		if s == nil {
+			return ""
+		}
+		return s.Type
+	}
+	return walkResources(p, func(r tfbridge.Resource) error {
+		id, hasID := r.TF.Schema().GetOk("id")
+		ok := hasID &&
+			(id.Type() == shim.TypeString || getIDType(r.Schema) == "string") &&
+			id.Computed()
+		if !ok {
+			r.Schema.ComputeID = missingID
+		}
+		return nil
+	})
+}
+
+func missingID(context.Context, resource.PropertyMap) (resource.ID, error) {
+	return "missing ID", nil
+}
+
+func fixPropertyConflict(p *info.Provider) error {
+	return walkResources(p, func(r tfbridge.Resource) error {
+		var retError error
+		r.TF.Schema().Range(func(key string, value shim.Schema) bool {
+			if fix := badPropertyName(key); fix != nil {
+				if r.Schema.Fields == nil {
+					r.Schema.Fields = map[string]*info.Schema{}
+				}
+
+				s, ok := r.Schema.Fields[key]
+				if !ok {
+					s = &info.Schema{}
+				}
+
+				if s.Name == "" {
+					var err error
+					s.Name, err = fix(r.TF)
+					if err != nil {
+						retError = fmt.Errorf("%q: %w", key, err)
+					}
+				}
+
+				if !ok {
+					r.Schema.Fields[key] = s
+				}
+			}
+
+			return true
+		})
+		return retError
+	})
+}
+
+func badPropertyName(key string) func(shim.Resource) (string, error) {
+	switch key {
+	case "urn":
+		return newName(urnReplaceList)
+	default:
+		return nil
+	}
+}
+
+var urnReplaceList = []string{
+	"resource_urn",
+}
+
+func newName(replaceList []string) func(shim.Resource) (string, error) {
+	return func(r shim.Resource) (string, error) {
+		s := r.Schema()
+		for _, v := range replaceList {
+			if _, ok := s.GetOk(v); !ok {
+				return tfbridge.TerraformToPulumiNameV2(v, s, nil), nil
+			}
+		}
+		return "", fmt.Errorf("no available new name, tried %#v", replaceList)
+	}
+}
+
+func walkResources(p *info.Provider, f func(tfbridge.Resource) error) error {
+	var errs []error
+
+	if p.Resources == nil {
+		p.Resources = map[string]*info.Resource{}
+	}
+
+	p.P.ResourcesMap().Range(func(key string, tf shim.Resource) bool {
+		res, isPresent := p.Resources[key]
+		if !isPresent {
+			res = &info.Resource{}
+		}
+		if err := f(tfbridge.Resource{
+			Schema: res,
+			TF:     tf,
+			TFName: key,
+		}); err != nil {
+			errs = append(errs, fmt.Errorf("%s: %w", key, err))
+		}
+
+		if !isPresent && !reflect.ValueOf(*res).IsZero() {
```

Comment:
The user isn't allowed to edit resources not present in the original list?
Why do we walk them in the first place, should we just skip them instead?

Reply:
The caller can edit resources not present in the original list. We don't need to re-insert them, since they are already in the original list. I'll add a comment explaining that logic.